src/envs/pybullet_blocks.py

```diff
@@ -0,0 +1,728 @@
+        if CFG.pybullet_draw_debug:  # pragma: no cover
+            # Draw the workspace on the table for clarity.
+            p.addUserDebugLine([self.x_lb, self.y_lb, self.table_height],
+                               [self.x_ub, self.y_lb, self.table_height],
+                               [1.0, 0.0, 0.0],
+                               lineWidth=5.0)
+            p.addUserDebugLine([self.x_lb, self.y_ub, self.table_height],
+                               [self.x_ub, self.y_ub, self.table_height],
+                               [1.0, 0.0, 0.0],
+                               lineWidth=5.0)
+            p.addUserDebugLine([self.x_lb, self.y_lb, self.table_height],
+                               [self.x_lb, self.y_ub, self.table_height],
+                               [1.0, 0.0, 0.0],
+                               lineWidth=5.0)
+            p.addUserDebugLine([self.x_ub, self.y_lb, self.table_height],
+                               [self.x_ub, self.y_ub, self.table_height],
+                               [1.0, 0.0, 0.0],
+                               lineWidth=5.0)
+            # Draw coordinate frame labels for reference.
+            p.addUserDebugText("x", [0.25, 0, 0], [0.0, 0.0, 0.0])
+            p.addUserDebugText("y", [0, 0.25, 0], [0.0, 0.0, 0.0])
+            p.addUserDebugText("z", [0, 0, 0.25], [0.0, 0.0, 0.0])
+            # Draw the pick z location at the x/y midpoint.
+            mid_x = (self.x_ub + self.x_lb) / 2
+            mid_y = (self.y_ub + self.y_lb) / 2
+            p.addUserDebugText("*", [mid_x, mid_y, self.pick_z],
+                               [1.0, 0.0, 0.0])
+
+        # Set gravity.
+        p.setGravity(0., 0., -10., physicsClientId=self._physics_client_id)
+
+        # Determine the initial joint values.
+        self._robot_home_pose = (self.robot_init_x, self.robot_init_y,
+                                 self.robot_init_z)
+        self._initial_joint_values = inverse_kinematics(
+            self._fetch_id,
+            self._ee_id,
+            self._robot_home_pose,
+            self._ee_orientation,
+            self._arm_joints,
+            physics_client_id=self._physics_client_id)
+
+        # Create blocks. Note that we create the maximum number once, and then
+        # remove blocks from view based on the number involved in the state.
+        num_blocks = max(max(CFG.blocks_num_blocks_train),
+                         max(CFG.blocks_num_blocks_test))
+        self._block_ids = [self._create_block(i) for i in range(num_blocks)]
+        self._block_id_to_block: Dict[int, Object] = {}
+
+        # When a block is held, a constraint is created to prevent slippage.
+        self._held_constraint_id: Optional[int] = None
+        self._held_block_id: Optional[int] = None
+
+    @property
+    def action_space(self) -> Box:
+        # dimensions: [dx, dy, dz, dfingers]
+        return Box(low=-0.05, high=0.05, shape=(4, ), dtype=np.float32)
```

Comment:
i self-answered below

Reply:
I didn't actually try anything larger (just used what we had before). I think you're right about the precision, and relatedly, we want to keep the velocity near 0 because velocity isn't modeled in the state.